kuka_rsi_hw_interface/src/rsi_command_handler.cpp

```diff
@@ -0,0 +1,325 @@
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    decodeNode(state_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+bool RSICommandHandler::Encode(char * & buffer, const size_t buffer_size)
+{
+  std::cout << "Decode started" << std::endl;
+  auto buffer_it = buffer;
+  try {
+    encodeNode(command_data_structure_, buffer, buffer_it, buffer_size);
+  } catch (const std::exception & e) {
+    std::cout << "### ERROR ###" << std::endl;
+    std::cerr << e.what() << '\n';
+    return false;
+  }
+  std::cout << "Decode finished" << std::endl;
+  return true;
+}
+
+void RSICommandHandler::decodeNode(
+  xml::XMLElement & element, char * const buffer, char * & buffer_it,
+  const size_t buffer_size)
+{
+  xml::XMLString node_name(buffer, 0);
+  // fast forward for the first open bracket
+  for (; *buffer_it != '<'; buffer_it++) {
+    if ((int)(buffer_it - buffer) >= (int)buffer_size) {
+      throw std::range_error("Out of the buffers range");
+    }
+  }
+  buffer_it++;
+  // Validate nodes name
+  if (!element.IsNameValid(node_name, buffer_it)) {
+    auto i = std::sprintf(err_buf_, "The detected name (");
+    std::snprintf(err_buf_ + i, node_name.length_ + 1, "%s", node_name.data_ptr_);
+    i += node_name.length_;
+    std::sprintf(
+      err_buf_ + i, ") does not match the elements name %s.",
+      element.GetName().c_str());
+    throw std::logic_error(err_buf_);
+  }
+
+  // Validate nodes params
+  size_t numOfParam = 0;
+  bool isBaseLessNode = false;
+  bool isNoMoreParam = false;
+  xml::XMLString node_param;
+  while (!isNoMoreParam && (size_t)(buffer_it - buffer) < buffer_size) {
+    // fast forward to the next non-space character
+    for (; *buffer_it == ' '; buffer_it++) {
+    }
+    if ((size_t)(buffer_it - buffer) >= buffer_size) {
+      throw std::range_error("Out of the buffers range");
+    }
+    // Check for the nodes hadders end characters
+    if (*buffer_it == '/' && *(buffer_it + 1) == '>') {
+      isBaseLessNode = true;
+      isNoMoreParam = true;
+    } else if (*buffer_it == '>') {
+      isNoMoreParam = true;
+    } else {
+      if (!element.IsParamNameValid(node_param, buffer_it)) {
+        auto i = std::sprintf(err_buf_, "The detected parameter (");
+        std::snprintf(err_buf_ + i, node_param.length_ + 1, "%s", node_param.data_ptr_);
+        i += node_param.length_;
+        std::sprintf(
+          err_buf_ + i, ") does not match with any of the %s elements parameters.",
+          element.GetName().c_str());
+        throw std::logic_error(err_buf_);
+      }
+      if (*buffer_it != '=') {
+        auto i = sprintf(err_buf_, "In \"");
+        std::snprintf(err_buf_ + i, node_param.length_ + 1, "%s", node_param.data_ptr_);
```

Comment:
why the + 1 here?

Reply:
it is +i, needed to increase the buffer pointer because fo the "In \""

Reply:
I know, I was referring to the node_param.length_ + 1 instead of node_param.length_

Reply:
resolved